renku/ui/service/gateways/repository_cache.py

```diff
@@ -0,0 +1,233 @@
+        """Evict expired projects from cache."""
+        for project in Project.all():
+            if project.ttl_expired():
+                self.evict(project)
+
+    def _update_project_access_date(self, project: Project):
+        """Update the access date of the project to current datetime."""
+        project.accessed_at = datetime.utcnow()
+        project.save()
+
+    def _clone_project(
+        self, cache: ServiceCache, git_url: str, branch: Optional[str], user: User, shallow: bool = True
+    ) -> Project:
+        """Clone a project to cache."""
+        try:
+            parsed_git_url = GitURL.parse(git_url)
+        except UnicodeError as e:
+            raise ValidationError("`git_url` contains unsupported characters", "git_url") from e
+        except errors.InvalidGitURL as e:
+            raise ValidationError("Invalid `git_url`", "git_url") from e
+
+        if parsed_git_url.owner is None or parsed_git_url.name is None:
+            raise ValidationError("Invalid `git_url`, missing owner or repository", "git_url")
+
+        project_data = {
+            "project_id": uuid.uuid4().hex,
+            "owner": parsed_git_url.owner,
+            "name": parsed_git_url.name,
+            "slug": normalize_to_ascii(parsed_git_url.name),
+            "depth": PROJECT_CLONE_DEPTH_DEFAULT if shallow else None,
+            "branch": branch,
+            "git_url": git_url,
+            "user_id": user.user_id,
+        }
+        project = cache.make_project(user, project_data, persist=False)
+
+        # NOTE: Create parent dir so lock file can be created.
+        project.abs_path.parent.mkdir(parents=True, exist_ok=True)
+
+        try:
+            with project.write_lock(), renku_project_context(project.abs_path, check_git_path=False):
+                try:
+                    # NOTE: If two requests ran at the same time, by the time we acquire the lock a project might
+                    # already be cloned by an earlier request.
+                    found_project = Project.get(
+                        (Project.user_id == user.user_id)
+                        & (Project.git_url == git_url)
+                        & (Project.branch == branch)
+                        & (Project.project_id != project.project_id)
+                    )
+                except ValueError:
+                    pass
+                else:
+                    if found_project.abs_path.exists():
+                        service_log.debug(f"project already cloned, skipping clone: {git_url}")
+                        self._update_project_access_date(found_project)
+                        return found_project
+
+                # clean directory in case of previous failed state
+                for root, dirs, files in os.walk(project.abs_path):
+                    for f in files:
+                        os.unlink(os.path.join(root, f))
+                    for d in dirs:
+                        shutil.rmtree(os.path.join(root, d))
```

Comment:
Why not `shutil.rmtree` on the root folder?

Reply:
We don't want to delete the root folder, only its contents. And we can't recreate the root folder afterwards, otherwise file handles pointing to the previous one  (e.g. the lock) will fail.